File: problems/ruby/030721-binary-tree-inorder-traversal/code.rb
# Definition for a binary tree node.
class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end


# https://leetcode.com/problems/binary-tree-inorder-traversal/
#
# @param {TreeNode} root
# @return {Integer[]}

def inorder_traversal(root)
  return [] if root.nil?

  inorder_traversal(root.left) + [root.val] + inorder_traversal(root.right)
end

File: problems/javascript/18072021-btree-preorder-traversal/code.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * https://leetcode.com/problems/binary-tree-preorder-traversal/
 *
 * @param {TreeNode} root
 * @return {number[]}
 */

const preorderTraversal = (root) => {
  if (!root) return []

  return [root.val]
    .concat(preorderTraversal(root.left))
    .concat(preorderTraversal(root.right))
}
